File: Algo/LeadToOffer/74_ProbOfDiceNum.cpp
// dp. O(n) space, O(n^2) time.
class Solution {
public:
    vector<double> dicesProbability(int n) {
        vector<double> prob(6, 1.0/6.0);
        if(n==1) return prob;

        for(int i=2; i<=n; i++){
            vector<double> tmp(5*i+1); // From i~6i
            for(int k=0; k<prob.size(); k++){
                for(int p=0; p<6; p++){
                    tmp[p+k] += prob[k] / 6.0;
                }
            }
            prob = tmp;
        }
        return prob;

    }
};

File: Algo/LeadToOffer/74_ProbOfDiceNum.py
class Solution(object):
    def dicesProbability(self, n):
        """
        :type n: int
        :rtype: List[float]
        """
        res = [1.0 / 6.0] * 6
        while n > 1:
            s = len(res)    # for small to large
            new_res = [0] * (s + 5)

            for i, p in enumerate(res):
                for k in range(6):
                    new_res[i+k] += res[i] * 1.0/6.0

            res = new_res
            n -= 1

        return res